core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -95,22 +86,31 @@ public static Ec2MetadataAsyncClient.Builder builder() {
 
     @Override
     public CompletableFuture<MetadataResponse> get(String path) {
-        return get(path, RetryPolicyContext.builder().retriesAttempted(0).build());
+        CompletableFuture<MetadataResponse> returnFuture = new CompletableFuture<>();
+        get(path, RetryPolicyContext.builder().retriesAttempted(0).build(), returnFuture);
+        return returnFuture;
     }
 
-    private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext retryPolicyContext) {
+    private void get(String path, RetryPolicyContext retryPolicyContext, CompletableFuture<MetadataResponse> returnFuture) {
         SdkHttpFullRequest baseTokenRequest = requestMarshaller.createTokenRequest(tokenTtl);
+        CompletableFuture<String> tokenFuture = sendAsyncRequest(baseTokenRequest);
+        CompletableFutureUtils.forwardExceptionTo(returnFuture, tokenFuture);
+
+        CompletableFuture<MetadataResponse> result = tokenFuture.thenCompose(token -> {
+            SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
+            return sendAsyncRequest(baseMetadataRequest);
+        }).thenApply(MetadataResponse::create);
 
-        CompletableFuture<String> tokenRequest = sendAsyncRequest(baseTokenRequest);
-        CompletableFuture<MetadataResponse> result = tokenRequest
-            .thenCompose(token -> {
-                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
-                return sendAsyncRequest(baseMetadataRequest);
-            }).thenApply(MetadataResponse::create);
+        CompletableFutureUtils.forwardExceptionTo(tokenFuture, result);
 
-        return result.handle((response, error) -> {
-            if (response != null || !shouldRetry(retryPolicyContext, error)) {
-                return result;
+        result.whenComplete((response, error) -> {
+            if (response != null) {
+                returnFuture.complete(response);
+                return;
+            }
+            if (!shouldRetry(retryPolicyContext, error)) {
```

Comment:
Is it possible to have a case of response being null but there isn't an error?

Reply:
I don't think that this is possible, no